tests/test-product-receipt-validation.js

```diff
@@ -0,0 +1,281 @@
+  var receipt = makeReceipt();
+
+  beforeEach(function() {
+    helper.setUp();
+    helper.appSelf.origin = defaultProductUrl;
+    fxpay.configure({
+      appSelf: helper.appSelf,
+      receiptCheckSites: [
+        'https://fake-receipt-check-server.net',
+      ]
+    });
+  });
+
+  afterEach(function() {
+    helper.tearDown();
+    window.localStorage.clear();
+  });
+
+  it('succeeds for valid receipts', function(done) {
+    helper.appSelf.receipts = [receipt];
+
+    setUpReceiptCheck({
+      onValidationRequest: function(requestBody) {
+        assert.equal(requestBody, receipt);
+      },
+    });
+
+    fxpay.getProduct(helper.apiProduct.guid).then(function(product) {
+      assert.equal(product.productId, helper.apiProduct.guid);
+      assert.equal(product.name, helper.apiProduct.name);
+      assert.equal(product.smallImageUrl, helper.apiProduct.logo_url);
+      assert.typeOf(product.receiptInfo, 'object');
+      return product.validateReceipt();
+    }).then(function(product) {
+      assert.equal(product.receiptInfo.status, 'ok');
+      assert.equal(product.receiptInfo.receipt, receipt);
+      done();
+    }).catch(done);
+
+  });
+
+  it('finds receipts in local storage', function(done) {
+    window.localStorage.setItem(settings.localStorageKey,
+                                JSON.stringify([receipt]));
+
+    setUpReceiptCheck();
+
+    fxpay.getProduct(helper.apiProduct.guid).then(function(product) {
+      return product.validateReceipt();
+    }).then(function(product) {
+      assert.equal(product.productId, helper.apiProduct.guid);
+      done();
+    }).catch(done);
+
+  });
+
+  it('fails on validation errors', function(done) {
+    helper.appSelf.receipts = [receipt];
+    var receiptResponse = {status: "invalid", reason: "ERROR_DECODING"};
+
+    setUpReceiptCheck({
+      validatorResponse: receiptResponse,
+    });
+
+    fxpay.getProduct(helper.apiProduct.guid).then(function(product) {
+      return product.validateReceipt();
+    }).catch(function(error) {
+      assert.instanceOf(error, fxpay.errors.InvalidReceipt);
+      assert.equal(error.productInfo.productId, helper.apiProduct.guid);
+      assert.equal(error.productInfo.receiptInfo.status,
+                   receiptResponse.status);
+      assert.equal(error.productInfo.receiptInfo.reason,
+                   receiptResponse.reason);
+      done();
+    }).catch(done);
+
+  });
+
+  it('fails on validation service errors', function(done) {
+    helper.appSelf.receipts = [makeReceipt()];
+
+    setUpReceiptCheck({
+      onValidationResponse: function(request) {
+        request.respond(500, {}, 'Internal Error');
+      },
+    });
+
+    fxpay.getProduct(helper.apiProduct.guid).then(function(product) {
+      return product.validateReceipt();
+    }).catch(function(error) {
+      assert.instanceOf(error, fxpay.errors.BadApiResponse);
+      assert.typeOf(error.productInfo, 'object');
```

Comment:
Are there any properties of this object that should be tested for in the error case?